File: 2023/Days/Day3/day3.js
const fs = require("fs");

const reset = "\x1b[0m";
const red = "\x1b[31m";
const yellow = "\x1b[38;5;226m";

const input_filePath = "./Days/Day3/test.txt";
// const input_filePath = './Days/Day3/data.txt'

function run() {
  console.log("Day 3");
  const data = loadFile(input_filePath);

  if (data == null) {
    console.log("Failed reading file");
    return;
  }

  const result = [
    { "Part 1": partone(data) },
    // {'Part 2': parttwo(data)}
  ];

  displayResults(result);
}

function isDigit(character) {
  const numbers = ["0", "1", "2", "3", "4", "5", "6", "7", "8", "9"];
  return numbers.includes(character);
}

function isSymbol(character) {
  const symbolRegex = /[!@#$%^&*()_+\-=\[\]{};':"\\|,<>\/?]+/;
  return symbolRegex.test(character);
}

function partone(data) {
  console.log(yellow + "Part One" + reset);
  var result = 0;

  const lines = data.split('\n')

  const part_numbers = []

  for (const i in lines){
    const l = parseInt(i)
    const characters = lines[l].split('')
    const num = {
      n: '',
      startingIndex: -1
    }

    for (const c in characters){
      if (c == characters.length - 1){
        if (num.startingIndex != -1) {
          const starting_y = l - 1 < 0 ? l : l - 1
          const ending_y = l + 1 >= lines.length ? l : l + 1
          
          for (y = starting_y; y < ending_y; y++) {
            const starting_x = num.startingIndex - 1 < 0 ? num.startingIndex : num.startingIndex - 1
            const ending_x = starting_x + num.n.length >= characters.length ? starting_x + num.n.length - 1 : starting_x + num.n.length
            console.log(`SX: ${starting_x} | SY: ${starting_y} | ${ending_x} | ${ending_y}`)
            for (x = starting_x; x < ending_x; x++) {
              console.log(y,x)
            }
          }
          
          num.n = ''
          num.startingIndex = -1
          continue
        }
      }

      if (isDigit(characters[c])){
        if (num.startingIndex == -1){
          num.n = characters[c]
          num.startingIndex =  c
        }
        else {
          num.n += characters[c]
        }
      }
      else {
        if (num.startingIndex != -1){
          const starting_y = l - 1 < 0 ? l : l - 1
          const ending_y = l + 1 >= lines.length ? l : l + 1
          for (y = starting_y; y < ending_y; y++) {
            const starting_x = num.startingIndex - 1 < 0 ? num.startingIndex : num.startingIndex - 1
            const ending_x = starting_x + num.n.length >= characters.length ? starting_x + num.n.length - 1 : starting_x + num.n.length
            console.log(`SX: ${starting_x} | SY: ${starting_y} | ${ending_x} | ${ending_y}`)
            for (x = starting_x; x < ending_x; x++) {
              console.log(y,x)
            }
          }

          num.n = ''
          num.startingIndex = -1
        }
      }
    }
  }

  result = lines.length


  return result;
}

function parttwo(data) {
  console.log(yellow + "Part Two" + reset);
  const lines = data.split("\r\n");

  var result = 0;

  for (var l of lines) {
  }

  return result;
}

function displayResults(data) {
  console.log("The Results");
  console.log("-----------");
  console.log(data);
}

function loadFile(path) {
  process.stdout.write(`Reading File: ${path}... `);
  try {
    const data = fs.readFileSync(path, "utf8");
    console.log("Successful");
    return data;
  } catch (error) {
    console.log(`Failed\nError reading file ${path}. \n${error}`);
    return null;
  }
}

module.exports = {
  run: run,
};

File: 2023/Days/Day3/day3.py
with open('data.txt', 'r') as file:
    lines = file.readlines()

def isSymbol(character, use=0):
    if use == 1:
        return not character.isalpha() and not character.isdigit() and character != '.'
    return not character.isalpha() and not character.isdigit()

part_numbers = []
gears = []

for line_index,line in enumerate(lines):
    print(line_index, line)
    line = line.replace('\n', '')
    num = {
        'n': '',
        'startingValue': -1
    }

    for char_index, character in enumerate(line):
        # print(char_index, character)
               
        if character.isdigit():
            # print('Digit')
            if num['startingValue'] == -1:
                num['n'] = character
                num['startingValue'] = char_index
            else:
                num['n'] += character       
        elif isSymbol(character):
            if num['startingValue'] != -1:
                print(f"End of Num: {num['n']}")
                starting_y = line_index - 1 if line_index - 1 >= 0 else line_index
                ending_y = line_index + 1 if line_index + 1 < len(lines) else line_index

                starting_x = num['startingValue'] - 1 if num['startingValue'] - 1 >= 0 else num['startingValue']
                ending_x = num['startingValue'] + len(num['n']) if num['startingValue'] + len(num['n']) < len(line) else num['startingValue'] + len(num['n']) - 1
                
                print(f"SX: {starting_x} | SY: {starting_y} |EX: {ending_x} | EY: {ending_y} | Line: {line_index} | Num: {num['n']}")

                found = False;
                for y in range(starting_y, ending_y + 1):
                    for x in range(starting_x, ending_x + 1):
                        print(f"X: {x} | Y: {y} | SYM: {lines[y][x]}")
                        if isSymbol(lines[y][x], 1):
                            print(f"{lines[y][x]} found inside {x},{y}")
                            if lines[y][x] == '*':
                                gear_found = False
                                if gears:
                                    for gear in gears:
                                        if gear['x'] == x and gear['y'] == y:
                                            gear['gears'].append(num['n'])
                                            gear_found = True
                                    if not gear_found:
                                        gears.append({
                                            'x': x,
                                            'y': y,
                                            'gears': [num['n']]
                                        })
                                else:
                                    gears.append({
                                        'x': x,
                                        'y': y,
                                        'gears': [num['n']]
                                    })
                            found = True
                if found:
                    print("found")
                    part_numbers.append(num['n'])
                        
                num['n'] = ''
                num['startingValue'] = -1

        if char_index == len(line)-1:
            if num['startingValue'] != -1:
                print(f"End of Num: {num['n']}")
                starting_y = line_index - 1 if line_index - 1 >= 0 else line_index
                ending_y = line_index + 1 if line_index + 1 < len(lines) else line_index

                starting_x = num['startingValue'] - 1 if num['startingValue'] - 1 >= 0 else num['startingValue']
                ending_x = num['startingValue'] + len(num['n']) if num['startingValue'] + len(num['n']) < len(line) else num['startingValue'] + len(num['n']) - 1
                
                print(f"SX: {starting_x} | SY: {starting_y} |EX: {ending_x} | EY: {ending_y} | Line: {line_index} | Num: {num['n']}")

                found = False;
                for y in range(starting_y, ending_y + 1):
                    for x in range(starting_x, ending_x + 1):
                        print(f"X: {x} | Y: {y} | SYM: {lines[y][x]}")
                        if isSymbol(lines[y][x], 1):
                            print(f"{lines[y][x]} found inside {x},{y}")
                            if lines[y][x] == '*':
                                gear_found = False
                                if gears:
                                    for gear in gears:
                                        if gear['x'] == x and gear['y'] == y:
                                            gear['gears'].append(num['n'])
                                            gear_found = True
                                    if not gear_found:
                                        gears.append({
                                            'x': x,
                                            'y': y,
                                            'gears': [num['n']]
                                        })
                                else:
                                    gears.append({
                                        'x': x,
                                        'y': y,
                                        'gears': [num['n']]
                                    })
                            found = True
                if found:
                    part_numbers.append(num['n'])
                        
                num['n'] = ''
                num['startingValue'] = -1

result_pn = 0
for num in part_numbers:
    result_pn += int(num)

result_gr = 0
for gear in gears:
    if len(gear['gears']) == 2:
        ratio = 1
        for g in gear['gears']:
            ratio *= int(g)
        result_gr += ratio

# print(part_numbers.keys())
# print(gears)
print(result_gr)